ospfd/ospf_asbr.c

```diff
@@ -305,3 +307,895 @@ void ospf_redistribute_withdraw(struct ospf *ospf, uint8_t type,
+		 * But, it is configured not to advertise then
+		 * flush the originated external lsa.
+		 */
+		if (lsa)
+			ospf_external_lsa_flush(ospf, 0, &aggr->p, 0);
+		UNSET_FLAG(aggr->flags, OSPF_EXTERNAL_AGGRT_ORIGINATED);
+
+		if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+			zlog_debug(
+				"%s: Don't originate the summary address,It is configured to not-advertise.",
+				__func__);
+		return NULL;
+	}
+
+	/* Prepare the extrenal_info for aggregator */
+	memset(&ei_aggr, 0, sizeof(struct external_info));
+	ei_aggr.p = aggr->p;
+	ei_aggr.tag = aggr->tag;
+	ei_aggr.type = 0;
+	ei_aggr.instance = ospf->instance;
+	ei_aggr.route_map_set.metric = -1;
+	ei_aggr.route_map_set.metric_type = -1;
+
+	/* Summary route already originated,
+	 * So, Do nothing.
+	 */
+	if (CHECK_FLAG(aggr->flags, OSPF_EXTERNAL_AGGRT_ORIGINATED)) {
+		if (!lsa) {
+			flog_warn(EC_OSPF_LSA_MISSING,
+				  "%s: Could not refresh/originate %s/%d",
+				  __func__, inet_ntoa(aggr->p.prefix),
+				  aggr->p.prefixlen);
+			return NULL;
+		}
+
+		asel = (struct as_external_lsa *)lsa->data;
+		tag = (unsigned long)ntohl(asel->e[0].route_tag);
+
+		/* If tag modified , then re-originate the route
+		 * with modified tag details.
+		 */
+		if (tag != ei_aggr.tag) {
+			if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+				zlog_debug(
+					"%s: Route tag changed(old:%d new:%d,So refresh the summary route.(%s/%d)",
+					__func__, tag, ei_aggr.tag,
+					inet_ntoa(aggr->p.prefix),
+					aggr->p.prefixlen);
+
+			ospf_external_lsa_refresh(ospf, lsa, &ei_aggr,
+						  LSA_REFRESH_FORCE, 1);
+		}
+		return lsa;
+	}
+
+	if (lsa && IS_LSA_MAXAGE(lsa)) {
+		/* This is special case.
+		 * If a summary route need to be originated but where
+		 * summary route already exist in lsdb with maxage, then
+		 * it need to be refreshed.
+		 */
+		if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+			zlog_debug(
+				"%s: LSA is in MAX-AGE so refreshing LSA(%s/%d)",
+				__PRETTY_FUNCTION__, inet_ntoa(aggr->p.prefix),
+				aggr->p.prefixlen);
+
+		ospf_external_lsa_refresh(ospf, lsa, &ei_aggr,
+					  LSA_REFRESH_FORCE, 1);
+		SET_FLAG(aggr->flags, OSPF_EXTERNAL_AGGRT_ORIGINATED);
+		return lsa;
+	}
+
+	/* If the external route prefix same as aggregate route
+	 * and if external route is already originated as TYPE-5
+	 * then it need to be refreshed and originate bit should
+	 * be set.
+	 */
+	if (lsa && prefix_same((struct prefix *)&ei_aggr.p,
+			       (struct prefix *)&ei->p)) {
+		if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+			zlog_debug(
+				"%s: External route prefix is same as aggr so refreshing LSA(%s/%d)",
+				__PRETTY_FUNCTION__, inet_ntoa(aggr->p.prefix),
+				aggr->p.prefixlen);
+		ospf_external_lsa_refresh(ospf, lsa, &ei_aggr,
+					  LSA_REFRESH_FORCE, 1);
+		SET_FLAG(aggr->flags, OSPF_EXTERNAL_AGGRT_ORIGINATED);
+		return lsa;
+	}
+
+	if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+		zlog_debug("%s: Originate Summary route(%s/%d)", __func__,
+			   inet_ntoa(aggr->p.prefix), aggr->p.prefixlen);
```

Comment:
use `%pI4` to print ipv4 addresses rather than `inet_ntoa`

Reply:
done